modules/apps/transfer/types/packet.go

```diff
@@ -64,3 +68,119 @@ func (ftpd FungibleTokenPacketData) ValidateBasic() error {
 func (ftpd FungibleTokenPacketData) GetBytes() []byte {
 	return sdk.MustSortJSON(mustProtoMarshalJSON(&ftpd))
 }
+
+/*
+
+ADR-8 CallbackPacketData implementation
+
+FungibleTokenPacketData implements CallbackPacketData interface. This will allow middlewares targeting specific VMs
+to retrieve the desired callback addresses for the ICS20 packet on the source and destination chains.
+
+The Memo is used to ensure that the callback is desired by the user. This allows a user to send an ICS20 packet
+to a contract with ADR-8 enabled without automatically triggering the callback logic which may lead to unexpected
+behaviour.
+
+The Memo format is defined like so:
+
+```json
+{
+	// ... other memo fields we don't care about
+	"callback": {
+		"src_callback_address": {contractAddrOnSourceChain},
+		"dest_callback_address": {contractAddrOnDestChain},
+
+		// optional fields
+		"gas_limit": {intForCallback}
+	}
+}
```

Comment:
Do we need the top-level key? Can I suggest:

```json
{
"src_callback": {"address": "...",  "msg": ""},
"dest_callback": {"address": "...", "msg": ""},
}
```

That way the message can be strictly specified (it would only include the address and gas limit is it's expected). It also allows for different gas_limits on src and dest callbacks. 

Another alternative is to remove the "msg" key and just pass the whole content of "src_callback" when calling the callback, but I think this may generate issues if the received is expecting a specific message type that does not contain "address", "gas_limit", etc. 

(This is not a problem in practice in cosmwasm as serde will just ignore the fields not defined in the type, but it may be an issue if we later want to implement other receiver types)

Reply:
I think using a top level key is a good suggestion which I was considering previously, and will likely do. We've already removed the `"msg"`/`"custom_msg"` field already based on feedback from confio.

The thing is, these callbacks are not passed directly to the contract, [this interface](https://github.com/cosmos/ibc-go/blob/99e3ad79ee7c96ce9e588f6279c27eaed824d220/modules/apps/callbacks/types/expected_keepers.go) will be implemented independently in wasmd and wasmvm. A new entry point will be opened, and actors will not be able to send custom messages along with the callback.

It seemed to me that Confio will likely make this even more restrictive during implementation, such as allowing source callbacks only if the contract is the sender of the packet.

Reply:
Allowing ack callbacks only if the contract is the sender of the packet makes sense to me, that's what we do on ibc-hooks. 

IIUC, the idea of not parsing the message here is so that each implementation can do it as they please. I think that makes sense in theory, but we already are assuming that what is getting called-back is an addr (i.e.: a contract) so I don't think the assumption of "some information needs to be provided to that addr" is too much. 

Anyway, don't have a strong preference here, just feels a bit unnecessary to parse this twice for the same keys.